activerecord/lib/active_record/attributes.rb

```diff
@@ -245,7 +240,10 @@ def define_attribute(
 
       def load_schema! # :nodoc:
         super
-        attributes_to_define_after_schema_loads.each do |name, (type, options)|
+
+        deferred = ancestors.map { |klass| klass.try(:deferred_attribute_definitions) }.compact.reduce(&:reverse_merge)
```

Comment:
I find this line a bit hard to grok, what about merging more imperatively:

```suggestion
        deferred = ancestors.each_with_object({}) do |mod, memo|
          memo.reverse_merge!(mod.deferred_attribute_definitions) if mod.respond_to?(:deferred_attribute_definitions)
        end
```

Or walking up the superclass chain:

```patch
diff --git a/activerecord/lib/active_record/attributes.rb b/activerecord/lib/active_record/attributes.rb
index 962e15c1e8..88b250505b 100644
--- a/activerecord/lib/active_record/attributes.rb
+++ b/activerecord/lib/active_record/attributes.rb
@@ -240,17 +240,24 @@ def define_attribute(

       def load_schema! # :nodoc:
         super
+        define_deferred_attributes
+      end

-        deferred = ancestors.map { |klass| klass.try(:deferred_attribute_definitions) }.compact.reduce(&:reverse_merge)
+      def define_deferred_attributes(klass: self, defined: Set.new) # :nodoc:
+        deferred_attribute_definitions.each do |name, (type, options)|
+          next if defined.include?(name)

-        deferred.each do |name, (type, options)|
           if type.is_a?(Symbol)
-            adapter_name = ActiveRecord::Type.adapter_name_from(self)
+            adapter_name = ActiveRecord::Type.adapter_name_from(klass)
             type = ActiveRecord::Type.lookup(type, **options.except(:default), adapter: adapter_name)
           end

-          define_attribute(name, type, **options.slice(:default))
+          klass.define_attribute(name, type, **options.slice(:default))
+
+          defined.add(name)
         end
+
+        superclass.try(:define_deferred_attributes, klass: klass, defined: defined)
       end

       private
```

Reply:
If we use `try` here, we should require the core extension (`active_support/core_ext/object/try`).

Reply:
It might be a matter of preference though, I'd not prefer to change both setter side and using side (and drop exisiting `attributes_to_define_after_schema_loads` and introduce new one are too), unless it is obviously easier to understand.

How about a way to address it only in setter side (in `attribute`) like as #39473?